### PR TITLE
Capturing uncaught exceptions with raven

### PR DESCRIPTION
**Pull request by rzlee**
## what
capturing uncaught exceptions and logging them in this framework

## why
so we can bubble them up to sentry

## context
https://jira.sagansystems.com/browse/DEV-4147

### CODE REVIEW

**Comment by rzlee** [comment]
right now it captures uncaught exceptions, but the data does not make it into sentry

@gerad can you look at this?

File: logger.js [hunk 1]
```diff
@@ -34,8 +38,34 @@ function createLogger(serviceName) {
   log.error = function(message, meta) {
```
Comment (gerad) [inline logger.js]: 1. change this function to accept 3 arguments: `message, meta, error`

File: logger.js [hunk 2]
```diff
@@ -34,8 +38,34 @@ function createLogger(serviceName) {
...
     log.logger.log('error', message, meta);
```
Comment (gerad) [inline logger.js]: 2. when `log.error` is called, it should report to sentry via raven's [`captureException`](https://docs.getsentry.com/hosted/clients/node/)

File: logger.js [hunk 3]
```diff
@@ -34,8 +38,34 @@ function createLogger(serviceName) {
...
   };
+
+  log.logErrorAndExit = function(err) {
```
Comment (gerad) [inline logger.js]: add this:

```js
log.handleUncaughtException = function() {
  if (sentry) {
    client.patchGlobal((sentrySent, error) => {
      log.error('uncaught exception, exiting', { sentrySent }, error);
      exit(1);
    });
  } else {
     process.on('uncaughtException', (error) => {
       log.error('uncaught exception, exiting', null, error);
       exit();
    });
}
```

Comment (gerad) [inline logger.js]: Then in whatever code uses the library call `log.handleUncaughtException`

**Comment by rzlee** [comment]
updated. it now sends to sentry

File: logger.js [hunk 4]
```diff
@@ -31,11 +35,47 @@ function createLogger(serviceName) {
     log.logger.log('debug', message, meta);
   };
 
-  log.error = function(message, meta) {
+  log.error = function(message, meta, error) {
     log.logger.log('error', message, meta);
```
Comment (gerad) [inline logger.js]: we want to add the error to the meta and log it here

File: logger.js [hunk 5]
```diff
@@ -31,11 +35,47 @@ function createLogger(serviceName) {
...
+    if (client) {
+      client.captureException(error);
```
Comment (gerad) [inline logger.js]: we should send the meta information and the message to sentry as well… see their api documentation

File: logger.js [hunk 6]
```diff
@@ -31,11 +35,47 @@ function createLogger(serviceName) {
...
+    var meta = {};
+    if (err instanceof Error) {
+      meta.message = err.message;
+      meta.stack = err.stack;
+    } else {
+      meta.error = err;
+    }
+    log.error('uncaught exception, exiting', meta, err);
+    process.exit(1);
+  }
```
Comment (gerad) [inline logger.js]: this could probably be a private function

File: logger.js [hunk 7]
```diff
@@ -31,11 +35,47 @@ function createLogger(serviceName) {
...
   };
   log.logger = logger;
-  return log;
+
+  if (sentryDSN) {
+    client = new raven.Client(sentryDSN, {logger: serviceName});
+    log.debug('logging errors to sentry');
+  } else {
+    log.debug('not logging errors to sentry');
+  }
```
Comment (gerad) [inline logger.js]: nit: i _think_ i would move this to the top where you set the initial client and sentry dsn

File: README.md [hunk 8]
```diff
@@ -3,7 +3,8 @@
 Structured logging for Javascript
 
 ## Usage
-```var log = createLogger("your-service-name");```
+        var log = createLogger("your-service-name");
+        log.handleUncaughtException();
```
Comment (gerad) [inline README.md]: if you do

    ```js
    // multiline javascript comment
    var here = is.my.code();
    ```

you don't need indentation and you get syntax highlighting

Here's what it looks like

```js
// multiline javascript comment
var here = is.my.code();
```

File: logger.js (hunk 7 above)
Comment (rzlee) [inline logger.js]: i tried that but then log.debug is not defined yet

File: README.md [hunk 9]
```diff
@@ -15,7 +18,9 @@ Debug logging:
 
 Error logging:
 
-```log.error('message', optional-meta);```
+```log.error('message', optional-meta, error);```
```
Comment (gerad) [inline README.md]: nit: `optionalMeta`, `optionalError`

File: logger.js [hunk 10]
```diff
@@ -31,11 +35,46 @@ function createLogger(serviceName) {
...
-    log.logger.log('error', message, meta);
+  log.error = function(message, meta, error) {
+    log.logger.log('error', message, {meta, error});
+    if (client) {
+      client.captureException(error, {extra: {meta, message}});
+    }
+  };
+
+  function uncaughtException(meta, err) {
+    var meta = {};
```
Comment (gerad) [inline logger.js]: you're overwriting the passed in meta here

Comment (gerad) [inline logger.js]: `meta = meta || {};`

File: logger.js [hunk 11]
```diff
@@ -31,11 +35,46 @@ function createLogger(serviceName) {
...
+    if (err instanceof Error) {
+      meta.message = err.message;
+      meta.stack = err.stack;
```
Comment (gerad) [inline logger.js]: i'd do `meta.error = { message: err.message, stack: err.stack };`

File: logger.js [hunk 12]
```diff
@@ -31,11 +35,46 @@ function createLogger(serviceName) {
...
+    } else {
+      meta.error = err;
+    }
```
Comment (gerad) [inline logger.js]: and all of this code should be in `log.error`

**Pull request merged**